### ISSUE TITLE
Generative Perplexity reproduction

### ISSUE THREAD

**Issue by zelaki** [david3684/flm#6]
Hello Chanhyuk,

Congradulation for the paper, very interesting and exciting results. 

I downloaded the pretrained models you provided and tried to reproduce the Gen. PPL results in Table 10 for FLM trained on OWT.
However, sampling 1024 samples with 1024 steps with the provided evaluation script yields a different (much better) Gen. PPL than the one reported in the paper:
  - Gen. PPL: 36.03004455566406,
  - Entropy: 4.892283916473389,
  
Whereas in the paper you report 
   - Gen. PPL: 62.23
   - Entropy: 5.33

I am running the evaluation script as follows:

```bash
  python -m main \
      mode=sample_eval \
      seed=1 \
      model=small \
      model.length=1024 \
      data=openwebtext-split \
      algo=flm \
      eval.checkpoint_path=$CKPT_PATH \
      loader.batch_size=2 \
      mode=sample_eval \
      seed=1 \
      model=small \
      model.length=1024 \
      data=openwebtext-split \
      algo=flm \
      eval.checkpoint_path=$CKPT_PATH \
      loader.batch_size=2 \
      loader.eval_batch_size=32 \
      sampling.num_sample_batches=32 \
      sampling.steps=$STEPS \
      algo.double_temb=False \
      eval.disable_ema=False \
      +wandb.offline=true 
```

What could be causing that discrepancy?

Thank you very much in advance!
Theo

**Comment by zelaki** [comment]
Hi again.

I found that the issue was that I am using an older version of your codebase, where the softmax is applied directly on the model_output:

Old version:
```python
def _process_model_output(self, model_output, xt, sigma):
    del xt, sigma
    return model_output.log_softmax(dim=-1)
```

New Version: (with this I can reproduce your reported Gen. PPL)
```python
def _process_model_output(self, model_output, xt, sigma, cap_value = 30.0):
    del xt, sigma
    model_output = cap_value * torch.tanh(model_output / cap_value)
    return model_output.log_softmax(dim=-1)
```

Do you have a sense of how much this change affects training? 

Thanks again.

**Comment by david3684** [comment]
Hi!
Softcapping (both for attention and output) is added mostly for numerical stability of training, as we observed the training randomly explodes at some point due to very large logit values. However the overall performance of training is almost identical with or without it.

However, it should be matched for sampling if you used it during the training, because the output distribution changes. Current checkpoint i uploaded is trained with softcap so it should be matched.

It would don't affect the performance if you not use this, but you should be aware of stability. Actually I don't think the output capping (that you attached) is really necessary if applied to the attention.

One other possible solution is using QKNorm before attention instead of softcap, which also show identical performance but adds stability.

**Comment by zelaki** [comment]
Hi again, and apologies for the follow-up!

I wanted to ask about the Gen PPL numbers reported for MDLM. When I download the official checkpoint (linked in the MDLM repo), which is also trained for 1M iterations with a batch size of 512, and evaluate it on 1024 samples using the script from this repo, I get:

GenPPL: 44.16, Ent: 5.30

This differs noticeably from the values reported in Table 10 of your paper:

GenPPL: 105.15, Ent: 5.63

Could you clarify how these numbers were obtained?

**Comment by david3684** [comment]
This is due to low precision sampling on ancestral sampler, and the fix wasn't applied in the official code.
(Referring to same issue on Duo repo https://github.com/s-sahoo/duo/issues/16 )
I have fixed to full precision and it would recover the reported value. Please pull and retry.
